### PR TITLE
Improvements 1 [API-1209]

### PR DESCRIPTION
**Pull request by utku-caglayan**
- Clean global scope
- Refactor configuration bits and add tests for it
- Refactor scattered logging/printing to make it more centralized

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+	KeyPath            string
+	KeyPassword        string
+}
+
+type Config struct {
+	Hazelcast hazelcast.Config
+	SSL       SSLConfig
+}
+
+type PersistentFlags struct {
```
Comment (nevzatseferoglu) [inline config/config.go:64]: The name confuses with actual `PersistentFlag` of `cobra`. Also these are not a flag, they keeps the value of the flags and it makes them a lvalue. Therefore we can name as `RootPFlagValues` or something different.

File: commands/cluster/cluster.go [hunk 2]
```diff
@@ -17,35 +17,113 @@ package commands
...
+	}
+	cmd.AddCommand(
+		NewGetState(),
+		NewChangeState(),
+		NewShutdown(),
+		NewVersion())
+	return &cmd
+}
+
+func NewGetState() *cobra.Command {
```
Comment (nevzatseferoglu) [inline commands/cluster/cluster.go:53]: `NewGetState`, `NewShutdown` and `NewVersion` are almost identical. We can gather them in a good block of code ?

File: commands/cluster/cluster.go [hunk 3]
```diff
@@ -17,35 +17,113 @@ package commands
...
-		hzConf, err := internal.MakeConfig()
-		if err != nil {
-			return err
-		}
-		// check if it is cloud invocation
-		if hzConf.Cluster.Cloud.Token != "" {
-			fmt.Println(invocationOnCloudErrorMessage)
+const invocationOnCloudInfoMessage = "Cluster operations on cloud are not supported. Checkout https://github.com/hazelcast/hazelcast-cloud-cli for cluster management on cloud."
+
+func New() *cobra.Command {
```
Comment (nevzatseferoglu) [inline commands/cluster/cluster.go:31]: The definition of `New` explains too many context. So we can renamed it in a more precise way.

**Review by nevzatseferoglu — CHANGES_REQUESTED** [review]
Great improvements ❤️

File: main.go [hunk 4]
```diff
@@ -16,14 +16,58 @@
...
+	"github.com/hazelcast/hazelcast-commandline-client/config"
+	hzcerror "github.com/hazelcast/hazelcast-commandline-client/errors"
+)
+
+const (
+	exitOK    = 0
+	exitError = 1
 )
 
 func main() {
```
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply main.go:36]: Can you clarify? I did not get it.

File: main.go [hunk 5]
```diff
@@ -16,14 +16,58 @@
...
-	isInteractive := commands.IsInteractiveCall()
+	rootCmd, persistentFlags := commands.NewRoot()
+	rootCmd.SetErr(os.Stderr)
+	rootCmd.SetOut(os.Stdout)
+	// parse global persistent flags
+	subcmd, flags, err := rootCmd.Find(os.Args[1:])
```
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply main.go]: Thanks, fixed at https://github.com/hazelcast/hazelcast-commandline-client/pull/47/commits/e46365e1fa90cd22e9ffaeee4a8a349c1dac33d6

File: config/config.go (hunk 1 above)
Comment (utku-caglayan) [reply config/config.go]: Renamed it as `GlobalFlagValues` at https://github.com/hazelcast/hazelcast-commandline-client/pull/47/commits/e46365e1fa90cd22e9ffaeee4a8a349c1dac33d6

File: main.go [hunk 6]
```diff
@@ -16,14 +16,58 @@
...
+	err = subcmd.ParseFlags(flags)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(exitError)
```
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply main.go]: Thanks for the suggestions. I tried to do something similar to option-1 on https://github.com/hazelcast/hazelcast-commandline-client/pull/47/commits/b3c742b25725ff73e65b1c47cb5d7592fff5b373

**Comment by utku-caglayan** [comment]
Fixes #56

**Review by yuce — COMMENTED** [review]
A few ideas. I'll continue to review.

File: commands/cluster/cluster.go (hunk 3 above)
Comment (nevzatseferoglu) [reply commands/cluster/cluster.go:31]: `NewCmd` makes more sense I guess. Then the usage will be `cluster.NewCmd()`. WDYT ?

File: commands/cluster/cluster.go (hunk 2 above)
Comment (nevzatseferoglu) [reply commands/cluster/cluster.go:53]: I cannot see what constraints that we will be encountered when we create them as dynamic. But I'm the side of dynamic creation.

File: main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (nevzatseferoglu) [reply main.go]: Thank you. It is much more readable now.

File: main.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nevzatseferoglu) [reply main.go:31]: New main commit is resolving my comment on this topic.

File: commands/cluster/cluster.go [hunk 7]
```diff
@@ -17,35 +17,113 @@ package commands
...
-	ClusterCmd.AddCommand(clusterChangeStateCmd)
-	ClusterCmd.AddCommand(clusterShutdownCmd)
-	ClusterCmd.AddCommand(clusterVersionCmd)
+func NewVersion() *cobra.Command {
+	return &cobra.Command{
+		Use:   "version",
+		Short: "retrieve information from the cluster",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			defer internal.ErrorRecover()
+			conf := config.FromContext(cmd.Context())
```
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply commands/cluster/cluster.go:120]: Ways I can think of were:
- global variable or singleton with a getter (I would prefer if we don't go that way)
- closure (pass context to every NewCmd function and access it from closure scope)
- context

Can you think of any other alternative? Also would you prefer another one from this list?

File: commands/cluster/cluster.go (hunk 3 above)
Comment (yuce) [reply commands/cluster/cluster.go:31]: I think the problem is with the name of the package. If it is renamed to `clustercmd`, then `clustercmd.New` sounds natural.

Comment (yuce) [reply commands/cluster/cluster.go:31]: Also we may remove the `commands` folder and move `clustercmd`, etc. to project root.

File: commands/cluster/cluster.go (hunk 2 above)
Comment (utku-caglayan) [reply commands/cluster/cluster.go]: Thank you both for the feedback. Fixed at https://github.com/hazelcast/hazelcast-commandline-client/pull/47/commits/dd8bd57e2d0d1b3c924b63a24213a56b2a69dd5f as you suggested

File: commands/cluster/cluster.go [hunk 8]
```diff
@@ -17,35 +17,113 @@ package commands
...
+
+func New() *cobra.Command {
+	cmd := cobra.Command{
+		Use:   "cluster {get-state | change-state | shutdown | query} [--state new-state]",
+		Short: "administrative cluster operations",
+		Long:  `administrative cluster operations which controls a Hazelcast cluster by manipulating its state and other features`,
+		PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
+			conf := config.FromContext(cmd.Context())
+			// check if it is a cloud invocation
+			if conf.Cluster.Cloud.Token != "" {
```
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply commands/cluster/cluster.go]: Thanks for catching that, fixed at https://github.com/hazelcast/hazelcast-commandline-client/pull/47/commits/022bb8f1a3f1bbef1249fbe21f3c7f59b9ad01bc

File: commands/root.go [hunk 9]
```diff
@@ -16,134 +16,56 @@
...
 		Long:  "Hazelcast command-line client connects your command-line to a Hazelcast cluster",
 		Example: "`hzc map --name my-map put --key hello --value world` - put entry into map directly\n" +
 			"`hzc help` - print help",
-		RunE: func(cmd *cobra.Command, args []string) error {
-			if _, err := internal.MakeConfig(); err != nil {
-				return err
+		// Handle errors explicitly
+		SilenceErrors: true,
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			// Make sure command receive non-nil configuration
```
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply commands/root.go]: Fixed at https://github.com/hazelcast/hazelcast-commandline-client/pull/47/commits/022bb8f1a3f1bbef1249fbe21f3c7f59b9ad01bc

File: run.go [hunk 10]
```diff
@@ -120,3 +121,21 @@ func HandleError(err error) string {
...
+
+func RunCmd(ctx context.Context, root *cobra.Command) error {
+	ctx, cancel := context.WithCancel(ctx)
+	handleInterrupt(ctx, cancel)
+	return root.ExecuteContext(ctx)
+}
+
+func handleInterrupt(ctx context.Context, cancel context.CancelFunc) {
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, os.Kill)
```
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply run.go:120]: os.Interrupt is present as you highlighted. I couldn't understand your point

File: commands/cluster/cluster.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply commands/cluster/cluster.go]: We proceeded with closure approach after discussion

**Review by yuce — APPROVED** [review]
Looks good!

**Review by nevzatseferoglu — APPROVED** [review]
Looks great.

**Pull request merged**